src/manifolds/StiefelCanonicalMetric.jl

```diff
@@ -0,0 +1,153 @@
+of the vertical component. Then using the matrix exponential ``\operatorname{Exp}`` we introduce ``B`` and ``C`` as
+
+```math
+\begin{pmatrix}
+B\\C
+\end{pmatrix}
+\coloneqq
+\operatorname{Exp}\left(
+\begin{pmatrix}
+A & -R^{\mathrm{T}}\\ R & 0
+\end{pmatrix}
+\right)
+\begin{pmatrix}I_k\\0\end{pmatrix}
+```
+
+the exponential map reads
+
+```math
+q = \exp_p X = pC + QB.
+```
+For more details, see [^EdelmanAriasSmith1998][^Zimmermann2017].
+"""
+exp(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, ::Any...) where {n,k}
+
+function exp!(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, q, p, X) where {n,k}
+    A = p' * X
+    n == k && return mul!(q, p, exp(A))
+    QR = qr(X - p * A)
+    BC_ext = exp([A -QR.R'; QR.R 0*I])
+    q .= [p Matrix(QR.Q)] * @view(BC_ext[:, 1:k])
+    return q
+end
+
+@doc raw"""
+    inner(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, CanonicalMetric}, p, X, Y)
+
+compute the inner procuct on the [`Stiefel`](@ref) manifold with respect to the
+[`CanonicalMetric`](@ref). The formula reads
+
+```math
+g_p(X,Y) = \operatorname{tr}\bigl( X^{\mathrm{T}}(I_n - \frac{1}{2}pp^{\mathrm{T}})Y \bigr).
+```
+"""
+function inner(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, p, X, Y) where {n,k}
+    T = Base.promote_eltype(p, X, Y)
+    s = T(dot(X, Y))
+    if n != k
+        s -= dot(p'X, p'Y) / 2
+    end
+    return s
+end
+
+@doc raw"""
+    X = log(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, CanonicalMetric}, p, q; kwargs..)
+    log!(M::MetricManifold{ℝ, Stiefel{n,k,ℝ}, X, CanonicalMetric}, p, q; kwargs...)
+
+Compute the logarithmic map on the [`Stiefel`](@ref)`(n,k)` manifold with respect to the [`CanonicalMetric`](@ref)
+using a matrix-algebraic based approach to an iterative inversion of the formula of the
+[`exp`](@ref exp(::MetricManifold{ℝ, Stiefel{n,k,ℝ}, CanonicalMetric}, ::Any...) where {n,k}).
+
+The algorithm is derived in[^Zimmermann2017].
+
+# Keyword arguments
+
+Both `maxiter`(`=1e5`) and `tolerance`(`=1e-9`) can be given to the iterative process as
+stoppping criteria.
+
+[^Zimmermann2017]:
+    > Zimmermann, R.: _A matrix-algebraic algorithm for the Riemannian logarithm on the Stiefel manifold under the canoncial metric.
+    > SIAM Journal on Matrix Analysis and Applications 28(2), pp. 322-342, 2017.
+    > doi: [10.1137/16M1074485](https://doi.org/10.1137/16M1074485),
+    > arXiv: [1604.05054](https://arxiv.org/abs/1604.05054).
+"""
+log(::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric}, ::Any...) where {n,k}
+
+function log(
+    M::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric},
+    p,
+    q;
+    maxiter=1e5,
+    tolerance=1e-9,
+) where {n,k}
+    X = allocate_result(M, log, p, q)
+    log!(M, X, p, q; maxiter=maxiter, tolerance=tolerance)
+    return X
+end
+
+function log!(
+    ::MetricManifold{ℝ,Stiefel{n,k,ℝ},CanonicalMetric},
+    X,
+    p,
+    q;
+    maxiter=1e5,
+    tolerance=1e-9,
+) where {n,k}
+    M = p' * q
+    QR = qr(q - p * M)
+    V = Matrix(qr([M; QR.R]).Q * I)
```

Comment:
Same here, why `I`?

Reply:
This is due to the reduced form the QR would return, see https://github.com/JuliaLang/julia/issues/27397, at least without that, the Q factor is rectangular this way, it should be quadratic.

Reply:
Yes, confirmed it again, with `I` the matrix `V` is 2k x 2k (in our tests 4 x 4), otherwise only 2k x k (4x2).

Reply:
`Matrix(A)` creates a copy even if `A` is already a `Matrix` (which it should be after multiplying by `I`). Changing to `convert(Matrix, A)` avoids the copy unless `A` happens to not be a `Matrix`. I'll push the change as well.